File: src/test/java/ru/jerael/booktracker/backend/web/mapper/AuthWebMapperTest.java
package ru.jerael.booktracker.backend.web.mapper;

import org.junit.jupiter.api.Test;
import ru.jerael.booktracker.backend.domain.model.auth.*;
import ru.jerael.booktracker.backend.domain.model.verification.VerificationType;
import ru.jerael.booktracker.backend.web.dto.auth.*;
import java.time.Instant;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.assertEquals;

class AuthWebMapperTest {
    private final AuthWebMapper authWebMapper = new AuthWebMapper();

    private final UUID userId = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final String token = "123456";
    private final String accessToken = "access token";
    private final String refreshToken = "refresh token";
    private final String email = "test@example.com";
    private final String password = "password";

    @Test
    void toDomain_ConfirmRegistration() {
        ConfirmRegistrationRequest request = new ConfirmRegistrationRequest(userId, token);

        ConfirmRegistration domain = authWebMapper.toDomain(request);

        assertEquals(userId, domain.userId());
        assertEquals(token, domain.token());
    }

    @Test
    void toResponse() {
        TokenPair tokenPair = new TokenPair(accessToken, refreshToken);

        AuthResponse response = authWebMapper.toResponse(tokenPair);

        assertEquals(accessToken, response.accessToken());
        assertEquals(refreshToken, response.refreshToken());
    }

    @Test
    void toDomain_UserLogin() {
        LoginRequest request = new LoginRequest(email, password);

        UserLogin domain = authWebMapper.toDomain(request);

        assertEquals(email, domain.email());
        assertEquals(password, domain.password());
    }

    @Test
    void toDomain_RefreshTokenPayload() {
        RefreshTokensRequest request = new RefreshTokensRequest(refreshToken);

        RefreshTokenPayload domain = authWebMapper.toDomain(request);

        assertEquals(refreshToken, domain.refreshToken());
    }

    @Test
    void toDomain_LogoutPayload() {
        LogoutRequest request = new LogoutRequest(refreshToken);

        LogoutPayload domain = authWebMapper.toDomain(request);

        assertEquals(refreshToken, domain.refreshToken());
    }

    @Test
    void toDomain_ResendVerification() {
        ResendVerificationRequest request = new ResendVerificationRequest(userId, "REGISTRATION");

        ResendVerification domain = authWebMapper.toDomain(request);

        assertEquals(userId, domain.userId());
        assertEquals(VerificationType.REGISTRATION, domain.type());
    }

    @Test
    void toResponse_ResendVerificationResponse() {
        Instant expiresAt = Instant.now().plusSeconds(600);
        ResendVerificationResult result = new ResendVerificationResult(userId, email, expiresAt);

        ResendVerificationResponse domain = authWebMapper.toResponse(result);

        assertEquals(userId, domain.userId());
        assertEquals(email, domain.email());
        assertEquals(expiresAt, domain.expiresAt());
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/mapper/AuthWebMapperTest.kt
package ru.jerael.booktracker.backend.web.mapper

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import ru.jerael.booktracker.backend.factory.auth.AuthDomainFactory
import ru.jerael.booktracker.backend.factory.auth.AuthWebFactory

class AuthWebMapperTest {
    private val mapper = AuthWebMapper()
    
    @Test
    fun `toDomain should map ConfirmRegistrationRequest to ConfirmRegistration`() {
        val request = AuthWebFactory.createConfirmRegistrationRequest()
        
        val domain = mapper.toDomain(request)
        
        with(domain) {
            assertEquals(request.userId, userId)
            assertEquals(request.token, token)
        }
    }
    
    @Test
    fun `toResponse should map TokenPair to AuthResponse`() {
        val tokenPair = AuthDomainFactory.createTokenPair()
        
        val response = mapper.toResponse(tokenPair)
        
        with(response) {
            assertEquals(tokenPair.accessToken, accessToken)
            assertEquals(tokenPair.refreshToken, refreshToken)
        }
    }
    
    @Test
    fun `toDomain should map LoginRequest to UserLogin`() {
        val request = AuthWebFactory.createLoginRequest()
        
        val domain = mapper.toDomain(request)
        
        with(domain) {
            assertEquals(request.email, email)
            assertEquals(request.password, password)
        }
    }
    
    @Test
    fun `toDomain should map RefreshTokensRequest to RefreshTokenPayload`() {
        val request = AuthWebFactory.createRefreshTokensRequest()
        
        val domain = mapper.toDomain(request)
        
        with(domain) {
            assertEquals(request.refreshToken, refreshToken)
        }
    }
    
    @Test
    fun `toDomain should map LogoutRequest to LogoutPayload`() {
        val request = AuthWebFactory.createLogoutRequest()
        
        val domain = mapper.toDomain(request)
        
        with(domain) {
            assertEquals(request.refreshToken, refreshToken)
        }
    }
    
    @Test
    fun `toDomain should map ResendVerificationRequest to ResendVerification`() {
        val request = AuthWebFactory.createResendVerificationRequest()
        
        val domain = mapper.toDomain(request)
        
        with(domain) {
            assertEquals(request.userId, userId)
            assertEquals(request.type, type.name)
        }
    }
    
    @Test
    fun `toResponse should map ResendVerificationResult to ResendVerificationResponse`() {
        val result = AuthDomainFactory.createResendVerificationResult()
        
        val response = mapper.toResponse(result)
        
        with(response) {
            assertEquals(result.userId, userId)
            assertEquals(result.email, email)
            assertEquals(result.expiresAt, expiresAt)
        }
    }
}